### PR TITLE
Notification section

### PR DESCRIPTION
**Pull request by Oluwatosin12-bit**
This is an update on the realtime notification method, users (both vendors and customers) now get pop-up notifications for certain actions (like favoriting, booking, accepting, declining) as well as history of notifications. I also started out on the design layout.
Screenshot 2024-07-16 at 12 36 13 PM

https://www.loom.com/share/02a9644b32994d9fa6471e2c7cd3c608?sid=71360779-c4f0-495f-8a75-a94b03886f5c

### CODE REVIEW

File: FrontEnd/src/Appointment/AppointmentPage.css [hunk 1]
```diff
@@ -1,3 +1,161 @@
...
+    margin-top: 1rem;
+  }
+
+  .appointmentActions button {
+    width: 100%;
+  }
+}
+
+
+/* .appointmentTab {
```
Comment (pearlor) [inline FrontEnd/src/Appointment/AppointmentPage.css:134]: Are these commented out for later use?

File: FrontEnd/src/Appointment/AppointmentPage.jsx [hunk 2]
```diff
@@ -6,34 +6,56 @@ import {
...
+    type
   ) => {
     await acceptAppointment(customerID, vendorID, appointmentID);
-    userAppointmentChanges(customerID);
-    vendorAppointmentChanges(vendorID);
+    await socket.emit("sendNotification", {
+      userID: userData?.userID,
+      senderID: userData?.userID,
+      receiverID: appointment.customerUID,
+      senderName: userData?.UserName,
```
Comment (pearlor) [inline FrontEnd/src/Appointment/AppointmentPage.jsx:33]: Rename `UserName` to `userName` for consistency.

File: FrontEnd/src/Appointment/AppointmentPage.jsx [hunk 3]
```diff
@@ -6,34 +6,56 @@ import {
...
+    type
   ) => {
     await declineAppointment(customerID, vendorID, appointmentID);
-    userAppointmentChanges(customerID);
-    vendorAppointmentChanges(vendorID);
+    await socket.emit("sendNotification", {
+      userID: userData?.userID,
+      senderID: userData?.userID,
+      receiverID: appointment.customerUID,
+      senderName: userData?.UserName,
```
Comment (pearlor) [inline FrontEnd/src/Appointment/AppointmentPage.jsx:53]: Rename `UserName` to `userName` for consistency.

File: FrontEnd/src/Appointment/AppointmentPage.jsx [hunk 4]
```diff
@@ -75,67 +89,109 @@ function AppointmentPage({ userData }) {
     upcomingAppointmentData.length === 0
   ) {
     return (
-      <div className="appointmentPage">
-        <h2>No appointments found.</h2>
+      <div className={`appointmentPage ${theme}`}>
+        <div className="appointments">
```
Comment (pearlor) [inline FrontEnd/src/Appointment/AppointmentPage.jsx:93]: For this, probably better to have a different className like `appointmentsEmptyMessage` so if you want to change the styling for the empty state later on.

File: FrontEnd/src/Appointment/AppointmentPage.jsx [hunk 5]
```diff
@@ -75,67 +89,109 @@ function AppointmentPage({ userData }) {
...
+              <div className="appointmentActions">
+                <button
+                  className="appointmentButtons"
+                  onClick={() =>
+                    handleAcceptedAppointment(
+                      appointment,
+                      appointment.customerUID,
+                      appointment.vendorUID,
+                      appointment.docID,
+                      3
```
Comment (pearlor) [inline FrontEnd/src/Appointment/AppointmentPage.jsx:134]: Can this type be saved into a reusable const in your utils? Because it's not clear what 3 refers to.

File: FrontEnd/src/Appointment/AppointmentPage.jsx [hunk 6]
```diff
@@ -75,67 +89,109 @@ function AppointmentPage({ userData }) {
...
+                  Accept
+                </button>
+                <button
+                  className="appointmentButtons"
+                  onClick={() =>
+                    handleDeclinedAppointment(
+                      appointment.customerUID,
+                      appointment.vendorUID,
+                      appointment.docID,
+                      4
```
Comment (pearlor) [inline FrontEnd/src/Appointment/AppointmentPage.jsx:147]: Let's define the 4 into a const value in utils.

File: FrontEnd/src/UserAuthentication/LandingPage.jsx [hunk 7]
```diff
@@ -1,28 +1,106 @@
-import { Link } from "react-router-dom";
+import React, { useRef, useState } from "react";
+import { useTheme } from "../UseContext";
 import "./LandingPage.css";
 
 function LandingPage() {
+  const { theme, setTheme } = useTheme();
+  const [menuOpen, setMenuOpen] = useState(false);
```
Comment (pearlor) [inline FrontEnd/src/UserAuthentication/LandingPage.jsx:7]: Rename to `[isMenuOpen, setIsMenuOpen]`

File: FrontEnd/src/App.jsx [hunk 8]
```diff
@@ -26,35 +39,90 @@ function App() {
...
+  useEffect(() => {
+    if (socket) {
+      socket.on("firstEvent", (msg) => {
+      });
+    }
+  }, [socket]);
+
+
+  useEffect(() => {
+    if (socket && userUID) {
```
Comment (koshka) [inline FrontEnd/src/App.jsx:72]: As mentioned before, use explicit `null` check when using non-boolean values in `true/false` evaluation:

`userUID != null`

File: FrontEnd/src/App.jsx [hunk 9]
```diff
@@ -26,35 +39,90 @@ function App() {
...
-          path="/AppointmentPage"
-          element={<AppointmentPage userData={userData} />}
-        />
-      </Routes>
+      <ThemeProvider>
+        <Routes>
+          <Route path="/" element={<LandingPage socket={socket} />} />
+          <Route path="/EntryPage" element={<EntryPage socket={socket} />} />
+          <Route path="/LoginPage" element={<LoginPage />} />
+          <Route path="/SignUpPage" element={<SignUpPage theme={theme} />} />
```
Comment (koshka) [inline FrontEnd/src/App.jsx:85]: If you're using context for the theme, why do you still pass `theme` through props? It should come from `useContext` inside of these components

File: FrontEnd/src/Appointment/AppointmentPage.css (hunk 1 above)
Comment (koshka) [reply FrontEnd/src/Appointment/AppointmentPage.css:134]: As we've discussed previously, let's not commit any commented out code, even if it's gonna be used later.

File: FrontEnd/src/Appointment/AppointmentPage.css [hunk 10]
```diff
@@ -1,3 +1,161 @@
...
 }
+.appointmentPage {
+  margin-top: 100px;
+  background-color: var(--backgroundColor);
+  display: flex;
+  flex-direction: column;
+  flex: 1;
+  align-items: center;
+}
+.appointments{
```
Comment (koshka) [inline FrontEnd/src/Appointment/AppointmentPage.css:14]: Nit: space is missing before `{`. Should've been picked up by Prettier? 🤔

File: FrontEnd/src/Appointment/AppointmentPage.jsx [hunk 11]
```diff
@@ -75,67 +89,109 @@ function AppointmentPage({ userData }) {
...
-                    appointment.docID
-                  )
-                }
-              >
-                Accept
+    <div className={`appointmentPage ${theme}`}>
+      <div className="appointments">
+        <h2>Pending Appointments</h2>
+        {pendingAppointmentData.map((appointment, index) => (
+          <div key={index} className="appointmentTab">
```
Comment (koshka) [inline FrontEnd/src/Appointment/AppointmentPage.jsx:104]: Using `index` as React key is considered an [anti-pattern](https://legacy.reactjs.org/docs/lists-and-keys.html#:~:text=We%20don't%20recommend%20using,cause%20issues%20with%20component%20state.). Unique IDs are preferred.

File: FrontEnd/src/Appointment/AppointmentPage.jsx [hunk 12]
```diff
@@ -75,67 +89,109 @@ function AppointmentPage({ userData }) {
...
+                  Decline
+                </button>
+              </div>
+            )}
+          </div>
+        ))}
+
+        <h2>Upcoming Appointments</h2>
+        {upcomingAppointmentData.map((appointment, index) => (
+          <div key={index} className="appointmentTab">
```
Comment (koshka) [inline FrontEnd/src/Appointment/AppointmentPage.jsx:160]: Ditto about `key`

File: FrontEnd/src/BookingPage/BookingPage.jsx [hunk 13]
```diff
@@ -47,17 +42,20 @@ function BookingForm({ userData }) {
...
     }
   }, [appointmentData]);
-  const handleSubmit = async (event) => {
+
+  const handleSubmit = async (event, type) => {
     try {
       event.preventDefault();
       if (userUID === post.userId) {
-        setPopUpMessage("You cannot book an appointment with yourself");
+        setModalPopUpMessage("You cannot book an appointment with yourself");
```
Comment (koshka) [inline FrontEnd/src/BookingPage/BookingPage.jsx:58]: I think I've mentioned this in one of the previous PRs - can you just disable/remove "Book" option if a current user is looking at their own post, instead of showing this message? Seems like better UX.

File: FrontEnd/src/Header.css [hunk 14]
```diff
@@ -4,42 +4,76 @@
...
   text-decoration: none;
 }
 
 .navigationList li {
   margin-right: 10px;
+  color: var(--darkOne);
+  position: relative;
+}
+
+.counter{
```
Comment (koshka) [inline FrontEnd/src/Header.css:42]: Missing space before `{`. Code formatter didn't run?

File: FrontEnd/src/Header.css [hunk 15]
```diff
@@ -4,42 +4,76 @@
...
+  font-size: 12px;
+  display: flex;
+  align-items: center;
+  justify-content: center;
+  position: absolute;
+  top: -5px;
+  right: -5px;
+}
+
+.popNotifications{
```
Comment (koshka) [inline FrontEnd/src/Header.css:58]: Missing space before `{`

File: FrontEnd/src/Header.css [hunk 16]
```diff
@@ -4,42 +4,76 @@
...
-  font-weight: bold;
+
+.popNotification{
+  padding: 5px;
+  font-size: 14px;
 }
-.navigationList li a.activeLink {
-  color: red !important;
-  font-weight: bold !important;
+.notificationButton{
```
Comment (koshka) [inline FrontEnd/src/Header.css:73]: Missing space before `{`

File: FrontEnd/src/Header.jsx [hunk 17]
```diff
@@ -32,23 +54,43 @@ function Header({ userData }) {
...
           </li>
           <li>
             <NavLink to="/UserProfile">
               <i className="fa-solid fa-user profileIcon"></i>
             </NavLink>
           </li>
+          <li>
+            <button className="toggleBtn" onClick={handleThemeChange}>
+              <i
+                className={`far ${theme === "light" ? "fa-moon" : "fa-sun"}`}
```
Comment (koshka) [inline FrontEnd/src/Header.jsx:73]: Let's save theme values somewhere as an enum so they can be imported, instead of hardcoding string

File: FrontEnd/src/Header.jsx [hunk 18]
```diff
@@ -32,23 +54,43 @@ function Header({ userData }) {
...
+              ></i>
+            </button>
+          </li>
         </ul>
       </nav>
+      {openNotifications && notifications.length > 0 && (
+        <div className="popNotifications">
+          {notifications.map((notification, index) => (
+            <div key={index}>
```
Comment (koshka) [inline FrontEnd/src/Header.jsx:82]: As mentioned above, prefer use unique IDs instead of `index` for the keys

File: FrontEnd/src/HomePage/CategoryList.css [hunk 19]
```diff
@@ -1,13 +1,25 @@
...
 }
 
 .categories ul li {
   list-style: none;
-  margin-bottom: 10px;
-  color: var(--color6);
+  margin: 20px auto;
+  color: var(--darkOne);
+}
+.categories nav{
```
Comment (koshka) [inline FrontEnd/src/HomePage/CategoryList.css:19]: Space before `{`

File: FrontEnd/src/HomePage/CategoryList.css [hunk 20]
```diff
@@ -20,4 +32,8 @@
   .categories ul li {
     font-size: 10px;
   }
+
+  .categoriesList{
```
Comment (koshka) [inline FrontEnd/src/HomePage/CategoryList.css:36]: Ditto

File: FrontEnd/src/Modal.css [hunk 21]
```diff
@@ -8,6 +8,7 @@
   display: flex;
   justify-content: center;
   align-items: center;
+  z-index: 10001;
```
Comment (koshka) [inline FrontEnd/src/Modal.css:11]: Prefer to use the smallest possible `z-index` value, incrementally increasing when necessary, instead of very large numbers. In larger projects this becomes very hard to manage: https://www.codegram.com/blog/stop-ab-using-z-index/

File: FrontEnd/src/Modal.jsx [hunk 22]
```diff
@@ -1,6 +1,6 @@
 import "./Modal.css";
 function Modal({ isShown, onClose, children }) {
-  if (!isShown) {
+  if (isShown === false) {
```
Comment (koshka) [inline FrontEnd/src/Modal.jsx:3]: Is `isShown` type always non-nullable boolean? If so, we've discussed before that you can just use `if (isShown)`, as opposed to the case where there will be implicit an type cast, like `if(userID)` (preferred `if(userID != null)`).

File: FrontEnd/src/PostFullDisplay.css [hunk 23]
```diff
@@ -23,3 +23,7 @@
 .fullDisplayDetails {
   overflow: auto;
 }
+
+.hearted{
```
Comment (koshka) [inline FrontEnd/src/PostFullDisplay.css:27]: Space before `{`

File: FrontEnd/src/PostFullDisplay.jsx [hunk 24]
```diff
@@ -1,14 +1,31 @@
...
-function PostFullDisplay({ userUID, post, isShown, onClose }) {
+
+function PostFullDisplay({ userUID, post, isShown, onClose, userData, socket }) {
+  const [favorited, setFavorited] = useState(false)
   const navigate = useNavigate();
   const handleBookingFormOpen = () => {
     navigate("/BookingPage", { state: { post, userUID } });
   };
-  if (!isShown) {
+  if (isShown === false) {
```
Comment (koshka) [inline FrontEnd/src/PostFullDisplay.jsx:11]: Why did you change that? Is `isShown` nullable? If not, the previous version is better.

File: FrontEnd/src/PostFullDisplay.jsx [hunk 25]
```diff
@@ -34,7 +51,7 @@ function PostFullDisplay({ userUID, post, isShown, onClose }) {
               </p>
               <button onClick={handleBookingFormOpen}>Book Now</button>
               <span>
-                <i className="fa-regular fa-heart"></i>
+                {favorited ? (<i className="fa-solid fa-heart hearted" ></i>) : (<i className="fa-regular fa-heart" onClick={()=>handleNotification(1)}></i>)}
```
Comment (koshka) [inline FrontEnd/src/PostFullDisplay.jsx:54]: Nit: when there's no content, self-closing tag is preferred:
`<i className="fa-solid fa-heart hearted" />`

File: FrontEnd/src/UseContext.jsx [hunk 26]
```diff
@@ -0,0 +1,14 @@
+import { createContext, useContext, useState } from 'react';
+
+const ThemeContext = createContext();
+export const ThemeProvider = ({ children }) => {
+  const [theme, setTheme] = useState('light');
```
Comment (koshka) [inline FrontEnd/src/UseContext.jsx:5]: Define an enum for theme values

File: FrontEnd/src/UserAuthentication/AlternateSignUp.jsx [hunk 27]
```diff
@@ -0,0 +1,166 @@
+import { useNavigate, Link } from "react-router-dom";
+import { useState, useEffect, useRef } from "react";
+import { registerUser } from "./Auth";
+import { CATEGORIES } from "../Categories";
+import "./EntryPage.css";
+
+function SignUpPage({ theme }) {
```
Comment (koshka) [inline FrontEnd/src/UserAuthentication/AlternateSignUp.jsx:7]: Why is this called "alternate"?

File: FrontEnd/src/UserAuthentication/LandingPage.jsx [hunk 28]
```diff
@@ -1,28 +1,106 @@
...
+
+  const landingPageContentRef = useRef(null);
+
+  const handleThemeChange = () => {
+    setTheme(theme === "light" ? "dark" : "light");
```
Comment (koshka) [inline FrontEnd/src/UserAuthentication/LandingPage.jsx:12]: Same, use an enum

File: FrontEnd/src/UserAuthentication/SignUpPage.jsx [hunk 29]
```diff
@@ -1,15 +1,16 @@
...
+import { useNavigate} from "react-router-dom";
+import { useState, useEffect, useRef } from "react";
 import { registerUser } from "./Auth";
-import { CATEGORIES} from "../Categories";
-import "./SignUpPage.css";
+import { CATEGORIES } from "../Categories";
+import "./EntryPage.css";
 
-function SignUpPage() {
+function SignUpPage({ theme }) {
```
Comment (koshka) [inline FrontEnd/src/UserAuthentication/SignUpPage.jsx:7]: Use context to access the theme

File: FrontEnd/src/Appointment/AppointmentPage.jsx (hunk 2 above)
Comment (Oluwatosin12-bit) [reply FrontEnd/src/Appointment/AppointmentPage.jsx:33]: It's the name field in my database. It is so that I can differentiate them from form inputs

File: FrontEnd/src/UserAuthentication/LandingPage.jsx (hunk 28 above)
Comment (Oluwatosin12-bit) [reply FrontEnd/src/UserAuthentication/LandingPage.jsx:12]: This is the first page users see when they get the website URL, they wont be able to change the theme from the header

File: FrontEnd/src/UserAuthentication/AlternateSignUp.jsx (hunk 27 above)
Comment (Oluwatosin12-bit) [reply FrontEnd/src/UserAuthentication/AlternateSignUp.jsx:7]: On smaller screens, the signUp page stays different (it doesn't slide over like it does in bigger screens because of space)

File: FrontEnd/src/Appointment/AppointmentPage.jsx (hunk 11 above)
Comment (Oluwatosin12-bit) [reply FrontEnd/src/Appointment/AppointmentPage.jsx:104]: When I did this, I noticed repetition of posts in userFeed

File: FrontEnd/src/Appointment/AppointmentPage.css (hunk 1 above)
Comment (Oluwatosin12-bit) [reply FrontEnd/src/Appointment/AppointmentPage.css:134]: This was from a stashed commit I used for saving my work when my laptop was acting up. While backtracking on working commits, I must have selected it. It's been updated in my new PR though.

File: FrontEnd/src/UseContext.jsx (hunk 26 above)
Comment (Oluwatosin12-bit) [reply FrontEnd/src/UseContext.jsx:5]: Could you please elaborate on what you mean here?

File: FrontEnd/src/Appointment/AppointmentPage.jsx (hunk 11 above)
Comment (koshka) [reply FrontEnd/src/Appointment/AppointmentPage.jsx:104]: As long as post IDs are unique, there shouldn't be repetition. React would actually show a warning if there're duplicate keys.

File: FrontEnd/src/UseContext.jsx (hunk 26 above)
Comment (koshka) [reply FrontEnd/src/UseContext.jsx:5]: I meant let's not hardcode strings like `light`. Define all supported theme values as an enum or object and import from there.

File: FrontEnd/src/UserAuthentication/AlternateSignUp.jsx (hunk 27 above)
Comment (koshka) [reply FrontEnd/src/UserAuthentication/AlternateSignUp.jsx:7]: Maybe would be useful to have a comment at the top of the component explaining this, otherwise naming is a bit confusing.

File: FrontEnd/src/UserAuthentication/LandingPage.jsx (hunk 28 above)
Comment (koshka) [reply FrontEnd/src/UserAuthentication/LandingPage.jsx:12]: Same commend as above. What I mean is define theme values as a constant object/enum. Using hardcoded strings is a sure way to introduce bugs due to potential typos.
